pkg/registry/apis/provisioning/register.go

```diff
@@ -136,12 +164,91 @@ func (b *ProvisioningAPIBuilder) getRepository(ctx context.Context, obj runtime.
 	case provisioning.LocalRepositoryType:
 		return newLocalRepository(r, b.localFileResolver), nil
 	case provisioning.GithubRepositoryType:
-		return newGithubRepository(r), nil
+		return newGithubRepository(ctx, r), nil
 	case provisioning.S3RepositoryType:
 		return newS3Repository(r), nil
 	default:
+		return &unknownRepository{config: r}, nil
+	}
+}
+
+func (b *ProvisioningAPIBuilder) beginCreate(ctx context.Context, obj runtime.Object, opts *metav1.CreateOptions) (registry.FinishFunc, error) {
+	cfg, ok := obj.(*provisioning.Repository)
+	if !ok {
+		return nil, fmt.Errorf("object is *provisioning.Repository")
+	}
+
+	repo, err := b.getRepository(ctx, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := repo.OnCreate(ctx); err != nil {
+		return nil, err
+	}
+
+	return func(ctx context.Context, success bool) {
+		if !success {
+			// TODO log this.
+			return
+		}
+	}, nil
+}
+
+func (b *ProvisioningAPIBuilder) beginUpdate(ctx context.Context, obj, old runtime.Object, opts *metav1.UpdateOptions) (registry.FinishFunc, error) {
+	objCfg, ok := obj.(*provisioning.Repository)
+	if !ok {
+		return nil, fmt.Errorf("new object is *provisioning.Repository")
+	}
+	oldCfg, ok := old.(*provisioning.Repository)
+	if !ok {
+		return nil, fmt.Errorf("old object is *provisioning.Repository")
+	}
+
+	if oldCfg.Spec.Type != objCfg.Spec.Type {
+		return nil, fmt.Errorf("changing repository type is not supported")
+	}
+
+	repo, err := b.getRepository(ctx, objCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	oldRepo, err := b.getRepository(ctx, oldCfg)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
```suggestion
```
lets pass in the old config.

Reply:
offline discussion: we need to instantiate the oldRepo to call `deleteWebhook`.